File: FizzBuzz-TS/src/fizzbuzz.spec.ts
/** @flow */

import { bless, check, forall, generator, random, shrink } from 'jsverify';
import { every, range, size, split, zip } from 'lodash/fp';

import generateOutput, { BUZZ, divisibleBy3And5, FIZZ, FIZZ_BUZZ } from './fizzbuzz';

describe('FizzBuzzProperty', () => {
  const checkOptions = {
    quiet: true,
    tests: 26
  };

  it('Non empty', () => {
    const prop = check(
      forall(arbListLength(), listLength => {
        const output = generateOutput(listLength);

        return size(output) !== 0;
      }),
      checkOptions
    );

    expect(prop).toBe(true);
  });

  it(`Numbers only divisible by 3 are represented by ${FIZZ}`, () => {
    const prop = check(
      forall(arbListLength(), listLength => {
        const output = generateOutput(listLength);

        const onlyDivisibleBy3 = keepNumbersDivisibleBy(3)(5)(output);

        return every(([_, v]) => v === FIZZ)(onlyDivisibleBy3);
      }),
      checkOptions
    );

    expect(prop).toBe(true);
  });

  it(`Numbers only divisible by 5 are represented by ${BUZZ}`, () => {
    const prop = check(
      forall(arbListLength(), listLength => {
        const output = generateOutput(listLength);

        const onlyDivisibleBy5 = keepNumbersDivisibleBy(5)(3)(output);

        return every(([_, v]) => v === BUZZ)(onlyDivisibleBy5);
      }),
      checkOptions
    );

    expect(prop).toBe(true);
  });

  it(`Numbers divisible by 3 and 5 are represented by ${FIZZ_BUZZ}`, () => {
    const prop = check(
      forall(arbListLength(), listLength => {
        const output = generateOutput(listLength);

        const onlyDivisibleBy3And5 = keepNumbersDivisibleBy3And5(output);

        return every(valueIsFizzAndOrBuzz)(onlyDivisibleBy3And5);
      }),
      checkOptions
    );

    expect(prop).toBe(true);
  });
});

const arbListLength = () =>
  bless({
    generator: generator.bless(() => random(1, 50)),
    show: val => val,
    shrink: shrink.noop
  });

const valueIsFizzAndOrBuzz = ([value, output]) => {
  if (divisibleBy3And5(value)) {
    return output === FIZZ_BUZZ;
  } else {
    throw new Error(`Expected value to be 3 and/or 5 and got ${output}`);
  }
};

const keepNumbersDivisibleBy = (nbr: number) => (nbr2: number) => (nbrs: string) => {
  const outputWithIndexes = addIndex(split('\n')(nbrs));

  return outputWithIndexes.filter(([index, _]) => index % nbr === 0 && index % nbr2 !== 0);
};

const keepNumbersDivisibleBy3And5 = (nbrs: string) => {
  const outputWithIndexes = addIndex(split('\n')(nbrs));

  return outputWithIndexes.filter(([index, _]) => divisibleBy3And5(index));
};

type AddIndex = (str: string[]) => Array<[number, string]>;
const addIndex: AddIndex = (collection: string[]) =>
  zip(range(1, size(collection) + 1))(collection);

File: FizzBuzz-HS/test/FizzBuzzSpec.hs
module FizzBuzzSpec where
  
import FizzBuzz (generateOutput)
import Test.Hspec

{-# ANN spec "HLint: Redundunt do" #-}

main :: IO ()
main = hspec spec
  
spec :: Spec
spec =
  describe "FizzBuzzProperty" $ do
    it "Non empty" $ do
      length (generateOutput [1]) `shouldNotBe` 0

    it "FizzBuzz are numbers divisible divisible by 3 and 5" $ do
      filterNumbersOnlyDivisibleBy 15 (lines (generateOutput [1..15])) `shouldBe` ["FizzBuzz"]

    it "Fizz are numbers divisible divisible by 3" $ do
      filterNumbersOnlyDivisibleBy 3 (lines (generateOutput [1..6])) `shouldBe` ["Fizz","Fizz"]

    it "Fizz are numbers divisible divisible by 5" $ do
      filterNumbersOnlyDivisibleBy 5 (lines (generateOutput [1..10])) `shouldBe` ["Buzz","Buzz"]

filterNumbersOnlyDivisibleBy :: Int -> [String] -> [String]
filterNumbersOnlyDivisibleBy value = map(\(_,x) -> x) . filter (\(x,_) -> x `mod` value == 0) . zip [1..]


